### PR TITLE
light weight mask rcnn models

### PR DESCRIPTION
**Pull request by Jerryzcn**
* resnet18 mask rcnn 
* fpn resnet18 mask rcnn
* fpn bn resnet18 mask rcnn
* fpn bn mobilenet 1.0 mask rcnn

### CODE REVIEW

**Comment by mli** [comment]
Job PR-782-2 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-782/2/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/model_zoo/mask_rcnn/mask_rcnn.py [hunk 1]
```diff
@@ -540,3 +550,119 @@ def mask_rcnn_fpn_resnet101_v1d_coco(pretrained=False, pretrained_base=True, **k
...
+        name='fpn_resnet18_v1b', dataset='coco', pretrained=pretrained,
+        features=features, top_features=top_features, classes=classes,
+        box_features=box_features, mask_channels=256, rcnn_max_dets=1000,
+        short=800, max_size=1333, min_stage=2, max_stage=6,
+        train_patterns=train_patterns, nms_thresh=0.5, nms_topk=-1,
+        post_nms=-1, roi_mode='align', roi_size=(14, 14),
+        strides=(4, 8, 16, 32, 64), clip=4.42, rpn_channel=1024, base_size=16,
+        scales=(2, 4, 8, 16, 32), ratios=(0.5, 1, 2), alloc_size=(384, 384),
+        rpn_nms_thresh=0.7, rpn_train_pre_nms=12000, rpn_train_post_nms=2000,
+        rpn_test_pre_nms=6000, rpn_test_post_nms=1000, rpn_min_size=0,
```
Comment (hetong007) [inline gluoncv/model_zoo/mask_rcnn/mask_rcnn.py]: We need to discuss for a way to expose the `rpn_test_pre_nms` and `rpn_test_post_nms` params. Otherwise the small models won't be so fast anyway.

File: gluoncv/model_zoo/mask_rcnn/mask_rcnn.py [hunk 2]
```diff
@@ -538,5 +550,251 @@ def mask_rcnn_fpn_resnet101_v1d_coco(pretrained=False, pretrained_base=True, **k
         scales=(2, 4, 8, 16, 32), ratios=(0.5, 1, 2), alloc_size=(384, 384),
         rpn_nms_thresh=0.7, rpn_train_pre_nms=12000, rpn_train_post_nms=2000,
         rpn_test_pre_nms=6000, rpn_test_post_nms=1000, rpn_min_size=0,
-        num_sample=512, pos_iou_thresh=0.5, pos_ratio=0.25, deep_fcn=True,
+        num_sample=512, pos_iou_thresh=0.5, pos_ratio=0.25, target_roi_scale=2,
+        num_fcn_convs=4, **kwargs)
+
+
+def mask_rcnn_resnet18_v1b_coco(pretrained=False, pretrained_base=True, rpn_test_pre_nms=6000,
```
Comment (hetong007) [inline gluoncv/model_zoo/mask_rcnn/mask_rcnn.py]: @zhreshold any thoughts on this change of parameters? Should we do that for all rcnn-family models? Should we take a closer look at what other params should we align with `rpn_test_pre_nms ` and `rpn_test_post_nms`?

At least I see `rcnn_max_dets` should not be larger than `rpn_test_post_nms`. Maybe they should be the same.

Comment (zhreshold) [reply gluoncv/model_zoo/mask_rcnn/mask_rcnn.py]: I think we should take a look at all these pre_nms and post_nms arguments, some are not effective in one mode vs another (with or w/o RPN). Some arguments have dependencies as @hetong007 just mentioned, we need more proper sanify checks for these. @Jerryzcn

**Comment by mli** [comment]
Job PR-782-5 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-782/5/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-782-6 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-782/6/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/model_zoo/mask_rcnn/mask_rcnn.py [hunk 3]
```diff
@@ -158,21 +169,23 @@ class MaskRCNN(FasterRCNN):
         Names of categories, its length is ``num_class``.
     mask_channels : int, default is 256
         Number of channels in mask prediction
-    deep_fcn : boolean, default False
-            Whether to use deep mask branch (4 convs)
+    rcnn_max_dets : int, default is 1000
```
Comment (zhreshold) [inline gluoncv/model_zoo/mask_rcnn/mask_rcnn.py]: I think this one need a guard as well, I remember @hetong007  had a problem when setting `rpn_test_post_nms` to small number and it failed. Probably we can add it to unittest?

Comment (Jerryzcn) [reply gluoncv/model_zoo/mask_rcnn/mask_rcnn.py]: done

**Comment by zhreshold** [comment]
@hetong007 lgtm now, can you post your demo mask_rcnn config with small nms numbers so Jerry can add it to unittest?

**Comment by hetong007** [comment]
My demo calls
```python
mask_rcnn_resnet18_v1b_coco(pretrained=False, pretrained_base=True, rcnn_max_dets=30,
                            rpn_test_pre_nms=100, rpn_test_post_nms=30, **kwargs):
```

If keep ` rcnn_max_dets=1000`, an error occurs.

**Comment by mli** [comment]
Job PR-782-9 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-782/9/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
wait for pretrained model or merge first?

**Comment by hetong007** [comment]
Since the test has been passed, I think the pretrained weights already exist.

**Comment by mli** [comment]
Job PR-782-10 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-782/10/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-782-11 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-782/11/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@hetong007 No, the pre-trained models are not uploaded yet. They passed because if they are not in pretrained_list in model_zoo, they got initialized. Since it's merged, it't ok to open a follow up PR.

**Comment by hetong007** [comment]
Oh I see. Let's have a follow up PR to resolve that.
